text/0000-app-mixins.md

```diff
@@ -0,0 +1,130 @@
+
+- *mixin* - a named set of additions to a stack that can be used to make additive changes to the contract.
+- *application developer* - a person who is using buildpacks to transform their application code into an OCI image
+- *stack buildpack* - a type of buildpack that runs against the stack image(s) instead of an app. Among other things, it can provide mixins.
+
+An application developer may specify a list of mixins in their application's `project.toml` file like this:
+
+```toml
+[build]
+mixins = [ "<mixin name>" ]
+```
+
+When a command like `pack build` is run, the list of mixins will be processed before buildpacks are run. For each mixin name, the following will happen:
+
+* If the mixin name is prefixed with `build:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the build image.
+* If the mixin name is prefixed with `run:` (as per [RFC-0006](https://github.com/buildpacks/rfcs/blob/main/text/0006-stage-specific-mixins.md)) the mixin will be dynamically added to the run image.
+* If the mixin name does not have a prefix it will be dynamically added to both the build and run stack images.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When a list of mixins are required by buildpacks via the build plan and the build phase starts:
+
+1. The lifecycle will compare the list of mixins to those provided by the stack using the `io.buildpacks.stack.mixins` label.
+1. If all mixin names are provided by the stack, no further action is required.
+1. If any requested mixin is not provided by the stack, the lifecycle will run the detect phase for all stackpacks defined in the builder.
+1. The lifecycle will compare the mixins added to the build plan to see if they match they required mixins.
+1. If at least one stackpack passes detect and provides the required mixin(s), the lifecycle will execute the stackpack build phase for all passing stackpack(s). If no stackpacks pass detect, or no stackpacks provide the required mixin, the build will fail with a message that describes the mixins that could not be provided.
+1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.
+
+## Stack Buildpacks
+
+The complete details of a Stack buildpack are described in [another RFC proposal](https://github.com/buildpacks/rfcs/pull/111).
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.
```

Comment:
How would the list of mixins used be persisted from `pack build` to `pack rebase`? I believe that paketobuildpacks make sure that the application code (if it can be compiled into a binary) isn't on the final image - does this require that the `project.toml` file be kept? Or will the list of used mixins/stackpacks be persisted through another mechanism?

Reply:
This is addressed in [RFC-0069: Stack buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md#rebasing-an-app). They will be stored in a LABEL.